connector/src/main/java/org/geysermc/connector/network/translators/bedrock/entity/player/BedrockMovePlayerTranslator.java

```diff
@@ -59,26 +66,64 @@ public void translate(MovePlayerPacket packet, GeyserSession session) {
+            );
+
+            // head yaw, pitch, head yaw
+            Vector3f rotation = Vector3f.from(packet.getRotation().getY(), packet.getRotation().getX(), packet.getRotation().getY());
+            entity.setPosition(packet.getPosition().sub(0, EntityType.PLAYER.getOffset(), 0));
+            entity.setRotation(rotation);
+            entity.setOnGround(packet.isOnGround());
+
+            session.sendDownstreamPacket(playerRotationPacket);
+            return;
+        }
+
         // We need to parse the float as a string since casting a float to a double causes us to
         // lose precision and thus, causes players to get stuck when walking near walls
-        double javaY = packet.getPosition().getY() - EntityType.PLAYER.getOffset();
-        if (packet.isOnGround()) javaY = Math.ceil(javaY * 2) / 2;
+        double javaY = Double.parseDouble(Float.toString(packet.getPosition().getY())) - EntityType.PLAYER.getOffset();
 
         Vector3d position = Vector3d.from(Double.parseDouble(Float.toString(packet.getPosition().getX())), javaY,
                 Double.parseDouble(Float.toString(packet.getPosition().getZ())));
 
-        if(!session.confirmTeleport(position)){
+        if (!session.confirmTeleport(position)){
             return;
         }
 
+        if (session.getConnector().getConfig().isCacheChunks()) {
+            // With chunk caching, we can do some proper collision checks
+
+            session.updatePlayerBoundingBox(position);
+            session.correctPlayerPosition();
+
+            position = Vector3d.from(session.getPlayerBoundingBox().getMiddleX(), session.getPlayerBoundingBox().getMiddleY() - (session.getPlayerBoundingBox().getSizeY() / 2),
+                    session.getPlayerBoundingBox().getMiddleZ());
+        } else {
+            // When chunk caching is off, we have to rely on this
+            // It rounds the Y position up to the nearest 0.5
+            // This snaps players to snap to the top of stairs and slabs like on Java Edition
+            // However, it causes issues such as the player floating on carpets
+            if (packet.isOnGround()) javaY = Math.ceil(javaY * 2) / 2;
+            position = position.up(javaY - position.getY());
+        }
+
         if (!isValidMove(session, packet.getMode(), entity.getPosition(), packet.getPosition())) {
             session.getConnector().getLogger().debug("Recalculating position...");
             recalculatePosition(session, entity, entity.getPosition());
             return;
         }
 
         ClientPlayerPositionRotationPacket playerPositionRotationPacket = new ClientPlayerPositionRotationPacket(
-                packet.isOnGround(), position.getX(), position.getY(), position.getZ(), packet.getRotation().getY(), packet.getRotation().getX()
+                packet.isOnGround(), position.getX(),
+                Math.round(position.getY() * 10000.0D) / 10000.0d,
```

Comment:
What in the world is this for?

Reply:
It might be leftover debug code, or it might be to solve precision issues. I'll make sure everything still works without it.

Reply:
Fixed